File: ambeth/Ambeth.IoC/ambeth/util/DelegatingConversionHelper.cs
using System;
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Exceptions;

namespace De.Osthus.Ambeth.Util
{
    public class DelegatingConversionHelper : IConversionHelper, IInitializingBean, IDedicatedConverterExtendable
    {
        [LogInstance]
        public ILogger Log { private get; set; }

        public IConversionHelper DefaultConversionHelper { protected get; set; }

        protected bool dateTimeUTC;

        [Property(DateTimeUtil.DateTimeAsUTC, DefaultValue = "true")]
        public override bool DateTimeUTC
        {
            set
            {
                dateTimeUTC = value;
            }
        }

        protected readonly ClassTupleExtendableContainer<IDedicatedConverter> converters = new ClassTupleExtendableContainer<IDedicatedConverter>("dedicatedConverter", "type", true);

        public virtual void AfterPropertiesSet()
        {
            ParamChecker.AssertNotNull(DefaultConversionHelper, "DefaultConversionHelper");

            DefaultConversionHelper.DateTimeUTC = dateTimeUTC;
        }

        public override T ConvertValueToType<T>(Object value)
        {
            return (T)ConvertValueToType(typeof(T), value, null);
        }

        public override T ConvertValueToType<T>(Object value, Object additionalInformation)
        {
            return (T)ConvertValueToType(typeof(T), value, additionalInformation);
        }

        public override Object ConvertValueToType(Type expectedType, Object value)
        {
            return ConvertValueToType(expectedType, value, null);
        }

        public override Object ConvertValueToType(Type expectedType, Object value, Object additionalInformation)
        {
            if (expectedType == null)
            {
                return value;
            }
            if (value == null)
            {
                if (expectedType.IsValueType)
                {
                    return Activator.CreateInstance(expectedType);
                }
                return null;
            }
            if (expectedType.IsAssignableFrom(value.GetType()))
            {
                return value;
            }
            Object sourceValue = value;
            while (true)
            {
                Type sourceClass = sourceValue.GetType();
                IDedicatedConverter dedicatedConverter = converters.GetExtension(sourceClass, expectedType);
                if (dedicatedConverter == null)
                {
                    break;
                }
                Object targetValue;
                try
                {
                    targetValue = dedicatedConverter.ConvertValueToType(expectedType, sourceClass, sourceValue, additionalInformation);
                }
                catch (Exception e)
                {
                    throw RuntimeExceptionUtil.Mask(e, "Error occured while converting value: " + sourceValue);
                }
                if (targetValue == null)
                {
                    if (expectedType.IsValueType)
                    {
                        throw new Exception("It is not allowed that an instance of " + typeof(IDedicatedConverter).FullName + " returns null like "
                                + dedicatedConverter + " did for conversion from '" + sourceClass.FullName + "' to '" + expectedType + "'");
                    }
                    return null;
                }
                if (expectedType.IsAssignableFrom(targetValue.GetType()))
                {
                    return targetValue;
                }
                if (targetValue.GetType().Equals(sourceValue.GetType()))
                {
                    throw new Exception("It is not allowed that an instance of " + typeof(IDedicatedConverter).FullName
                            + " returns a value of the same type (" + targetValue.GetType().FullName + ") after conversion like " + dedicatedConverter
                            + " did");
                }
                sourceValue = targetValue;
            }
            if (expectedType.IsArray && sourceValue != null && sourceValue.GetType().IsArray)
            {
                // try to convert item by item of the array
                Array sourceArray = (Array)sourceValue;
                int size = sourceArray.GetLength(0);
                Type expectedComponentType = expectedType.GetElementType();
                Array targetValue = Array.CreateInstance(expectedComponentType, size);
                for (int a = sourceArray.GetLength(0); a-- > 0; )
                {
                    Object sourceItem = sourceArray.GetValue(a);
                    Object targetItem = ConvertValueToType(expectedComponentType, sourceItem);
                    targetValue.SetValue(targetItem, a);
                }
                return targetValue;
            }
            return DefaultConversionHelper.ConvertValueToType(expectedType, sourceValue, additionalInformation);
        }

        public void RegisterDedicatedConverter(IDedicatedConverter dedicatedConverter, Type sourceType, Type targetType)
        {
            converters.Register(dedicatedConverter, sourceType, targetType);
        }

        public void UnregisterDedicatedConverter(IDedicatedConverter dedicatedConverter, Type sourceType, Type targetType)
        {
            converters.Unregister(dedicatedConverter, sourceType, targetType);
        }
    }
}

File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/util/DelegatingConversionHelper.java
package de.osthus.ambeth.util;

import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;

public class DelegatingConversionHelper extends ClassTupleExtendableContainer<IDedicatedConverter> implements IInitializingBean, IDedicatedConverterExtendable,
		IConversionHelper
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	protected IConversionHelper defaultConversionHelper;

	public DelegatingConversionHelper()
	{
		super("dedicatedConverter", "type", true);
	}

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		ParamChecker.assertNotNull(defaultConversionHelper, "defaultConversionHelper");
	}

	public void setDefaultConversionHelper(IConversionHelper defaultConversionHelper)
	{
		this.defaultConversionHelper = defaultConversionHelper;
	}

	@Override
	public <T> T convertValueToType(Class<T> expectedType, Object value)
	{
		return convertValueToType(expectedType, value, null);
	}

	@SuppressWarnings("unchecked")
	@Override
	public <T> T convertValueToType(Class<T> expectedType, Object value, Object additionalInformation)
	{
		if (value == null)
		{
			return null;
		}
		if (expectedType == null || expectedType.isAssignableFrom(value.getClass()))
		{
			return (T) value;
		}
		if (expectedType.isPrimitive() && value instanceof Number)
		{
			return defaultConversionHelper.convertValueToType(expectedType, value);
		}
		try
		{
			Object targetValue = value;
			while (true)
			{
				Class<?> targetClass = targetValue.getClass();
				IDedicatedConverter dedicatedConverter = getExtension(targetClass, expectedType);
				if (dedicatedConverter == null)
				{
					break;
				}
				Object newTargetValue = dedicatedConverter.convertValueToType(expectedType, targetClass, targetValue, additionalInformation);
				if (newTargetValue == null)
				{
					if (expectedType.isPrimitive())
					{
						throw new IllegalStateException("It is not allowed that an instance of " + IDedicatedConverter.class.getName() + " returns null like "
								+ dedicatedConverter + " did for conversion from '" + targetClass.getName() + "' to '" + expectedType + "'");
					}
					return null;
				}
				if (expectedType.isAssignableFrom(newTargetValue.getClass()))
				{
					return (T) newTargetValue;
				}
				if (newTargetValue.getClass().equals(targetValue.getClass()))
				{
					throw new IllegalStateException("It is not allowed that an instance of " + IDedicatedConverter.class.getName()
							+ " returns a value of the same type (" + newTargetValue.getClass().getName() + ") after conversion like " + dedicatedConverter
							+ " did");
				}
				targetValue = newTargetValue;
			}
			return defaultConversionHelper.convertValueToType(expectedType, targetValue, additionalInformation);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	@Override
	public void registerDedicatedConverter(IDedicatedConverter dedicatedConverter, Class<?> sourceType, Class<?> targetType)
	{
		register(dedicatedConverter, sourceType, targetType);
	}

	@Override
	public void unregisterDedicatedConverter(IDedicatedConverter dedicatedConverter, Class<?> sourceType, Class<?> targetType)
	{
		unregister(dedicatedConverter, sourceType, targetType);
	}
}
